File: 2017/c/day01.c
#include <stdio.h>

#include "io-util.h"

void partA(INPUT *input) {
    int sum = 0;
    unsigned char *data = input->data;
    int prev = data[input->size - 1];
    for (int i = 0; i < input->size; i++) {
        int ch = data[i];
        if (ch == prev) {
            sum += ch - '0';
        }
        prev = ch;
    }
    printf("A: %d\n", sum);
}

void partB(INPUT *input) {
    int sum = 0;
    unsigned char *data = input->data;
    int delta = input->size >> 1;
    for (int i = 0; i < input->size; i++) {
        int ch = data[i];
        if (ch == data[(i + delta) % input->size]) {
            sum += ch - '0';
        }
    }
    printf("B: %d\n", sum);
}

int main() {
    INPUT *input = read_input("../input-01.txt", 1);
    if (input == NULL) {
        return 1;
    }

    partA(input);
    partB(input);

    free_input(input);
    return 0;
}

File: 2017/python/day01.py
with open('../input-01.txt') as fp:
    data = [x.strip() for x in fp.readlines()]

sum = 0
for line in data:
    prev = line[-1]
    for ch in line:
        if prev == ch:
            sum += int(ch)
        prev = ch
print("A: {}".format(sum))

sum = 0
for line in data:
    size = len(line)
    half = int(size/2)
    for i in range(size):
        if line[i] == line[(i+half) % size]:
            sum += int(line[i])
print("B: {}".format(sum))